### PR TITLE
Refactor GeoSortBuilder

### PR DESCRIPTION
**Pull request by MaineC**
- adds json parsing,
- refactors json serialisation,
- adds writable parsing and serialisation,
- adds json and writable roundtrip test

This relates to #15178

Caveat: Tests are still running locally, so expect changes to fix any failures to come.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 1]
```diff
@@ -22,33 +22,52 @@
...
-    private GeoDistance geoDistance;
-    private DistanceUnit unit;
-    private SortOrder order;
-    private String sortMode;
+    private GeoDistance geoDistance = GeoDistance.DEFAULT;
+    private DistanceUnit unit = DistanceUnit.DEFAULT;
+    private SortOrder order = SortOrder.DEFAULT;
+    
+    // TODO there is an enum that covers that parameter which we should be using here
+    private String sortMode = DEFAULT_SORT_MODE;
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Not sure if setting default to MAX here is the same as what GeoDistanceSortParser is doing at the moment. Look like if `sort_mode` isn't specified in json at the moment, it will be set to MIN or MAX depending on the `order`/`reverse` parameter. Maybe better leave this unassigned and resolve later if needed?

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 2]
```diff
@@ -60,6 +79,29 @@ public GeoDistanceSortBuilder(String fieldName) {
     }
 
     /**
+     * Copy constructor.
+     * */
+    public GeoDistanceSortBuilder(GeoDistanceSortBuilder original) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Maybe package private if only needed in the tests like at the moment? Or move there? No strong opinions though.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 3]
```diff
@@ -113,11 +181,18 @@ public GeoDistanceSortBuilder order(SortOrder order) {
...
+    }
+
     /**
      * Not relevant.
+     *
+     * TODO should this throw an exception rather than silently ignore a parameter that is not used?
      */
     @Override
-    public SortBuilder missing(Object missing) {
+    public GeoDistanceSortBuilder missing(Object missing) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Looks like `FieldSortBuilder` is the only impl of SortBuilder that actually supports `missing`, so I'd try to delete this from the interface and move it there.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 4]
```diff
@@ -147,21 +239,39 @@ public GeoDistanceSortBuilder setNestedPath(String nestedPath) {
...
+    public boolean ignoreMalformed() {
+        return this.ignoreMalformed;
+    }
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.startObject("_geo_distance");
-        if (geohashes.size() == 0 && points.size() == 0) {
+        builder.startObject(NAME);
+        if (points.size() == 0) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: This kind of validation feels odd in toXContent. Can we make point/point mandatory constructor args?

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 5]
```diff
@@ -169,23 +279,14 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
-        }
+        builder.field("unit", unit);
+        builder.field("distance_type", geoDistance.name().toLowerCase(Locale.ROOT));
         if (order == SortOrder.DESC) {
             builder.field("reverse", true);
         }
-        if (sortMode != null) {
-            builder.field("mode", sortMode);
-        }
+        builder.field("mode", sortMode);
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: See comment above, if sortMode can be `null`, this would need to be wrapped in check again.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 6]
```diff
@@ -193,14 +294,192 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
         return builder;
     }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public boolean equals(Object object) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Maybe add check for object reference equality.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 7]
```diff
@@ -193,14 +294,192 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
+        if (! (object instanceof GeoDistanceSortBuilder)) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: This made me wonder why we compare objects class in other places like this, maybe its faster? Was this a concious choice against it?

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 8]
```diff
@@ -22,33 +22,52 @@
...
-public class GeoDistanceSortBuilder extends SortBuilder {
+public class GeoDistanceSortBuilder extends SortBuilder implements ToXContent, NamedWriteable<GeoDistanceSortBuilder>, ParameterParser<GeoDistanceSortBuilder> {
+    public static final String NAME = "_geo_distance";
+    public static final boolean DEFAULT_COERCE = false;
+    public static final boolean DEFAULT_IGNORE_MALFORMED = false;
+    public static final String DEFAULT_SORT_MODE = "MAX"; 
+
+    static final GeoDistanceSortBuilder PROTOTYPE = new GeoDistanceSortBuilder("");
 
     final String fieldName;
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: can this be private?

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 9]
```diff
@@ -193,14 +294,192 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
+                } else {
+                    GeoPoint point = new GeoPoint();
+                    point.resetFromString(parser.text());
+                    geoPoints.add(point);
+                    fieldName = currentName;
+                }
+            }
+        }
+
+        if (sortMode == null) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: This cannot happen anymore when setting the default, but since this was the behaviour in the "old" style parser, I wonder if we need to keep it?

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 10]
```diff
@@ -193,14 +294,192 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentName = parser.currentName();
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                parseGeoPoints(parser, geoPoints);
+
+                fieldName = currentName;
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                // the json in the format of -> field : { lat : 30, lon : 12 }
+                if ("nested_filter".equals(currentName) || "nestedFilter".equals(currentName)) {
+                    nestedFilter = context.parseInnerQueryBuilder();
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Note to remember: while this is kept as a QueryBuilder internally, I think we need to make sure to call `toFiler()` on it once on the shard (e.g. in the new build() method, doesn't seem to be there yet)

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 11]
```diff
@@ -193,14 +294,192 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
+        result.sortMode(sortMode.name());
+        result.setNestedFilter(nestedFilter);
+        result.setNestedPath(nestedPath);
+        result.coerce(coerce);
+        result.ignoreMalformed(ignoreMalformed);
+        return result;
+
+    }
+
+    private void parseGeoPoints(XContentParser parser, List<GeoPoint> geoPoints) throws IOException {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Since this seems exact duplicate of code in GeoDistanceSortParser, and can be made static, can this be used from GeoDistanceSortParser or vice versa?

File: core/src/main/java/org/elasticsearch/search/sort/ParameterParser.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+package org.elasticsearch.search.sort;
+
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.index.query.QueryParseContext;
+
+import java.io.IOException;
+
+public interface ParameterParser<T extends ToXContent> {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/ParameterParser.java]: Do we need this interface? Currently I think it's fromXContent signature will only match the SortBuilders. Maybe just extend that interface instead?

File: core/src/test/java/org/elasticsearch/search/sort/AbstractSearchSourceItemTestCase.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+    static IndicesQueriesRegistry indicesQueriesRegistry;
+
+    @BeforeClass
+    public static void init() {
+        namedWriteableRegistry = new NamedWriteableRegistry();
+        namedWriteableRegistry.registerPrototype(MatchAllQueryBuilder.class, MatchAllQueryBuilder.PROTOTYPE);
+        namedWriteableRegistry.registerPrototype(IdsQueryBuilder.class, IdsQueryBuilder.PROTOTYPE);
+        namedWriteableRegistry.registerPrototype(TermQueryBuilder.class, TermQueryBuilder.PROTOTYPE);
+        namedWriteableRegistry.registerPrototype(FieldSortBuilder.class, FieldSortBuilder.PROTOTYPE);
+        namedWriteableRegistry.registerPrototype(GeoDistanceSortBuilder.class, GeoDistanceSortBuilder.PROTOTYPE);
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/sort/AbstractSearchSourceItemTestCase.java]: This doesn't compile for me, also Nik changed something with the setup of the query parsers in (15c43b64) , so maybe this is not needed any more. I think it should be enough to add the Sort-Prototype for tests here

File: core/src/test/java/org/elasticsearch/search/sort/AbstractSearchSourceItemTestCase.java [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+import org.junit.BeforeClass;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.hamcrest.Matchers.*;
+
+//TODO maybe merge with AbstractsortBuilderTestCase once #14933 is in?
+public abstract class AbstractSearchSourceItemTestCase<T extends NamedWriteable<T> & ToXContent & ParameterParser<T>> extends ESTestCase {
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/sort/AbstractSearchSourceItemTestCase.java]: This looks like a very generic test, but then it mentions sort in many comments and variable names. Maybe call it something sort-specific. Also I don't get the TODO.

**Comment by cbuescher** [comment]
@MaineC I wasn't sure if this was ready to review or you want to include the build method as well, but I went ahead and left a couple of minor comments. Already look very good to me besides the missing method that creates the SortField.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 1 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Makes sense.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 2 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Done.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 3 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Good catch. Moved to a separate PR here #16225

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 1 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: After going through the change and tests once more: Not setting this to some default, the toXContent/fromXContent roundtrip test fails - in fromXContent the value is being set, if it's set, it's being rendered in toXContent, even if in the initial builder it was neither set nor rendered out. Not sure how to solve this issue.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 4 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: done

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 6 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: +1

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 7 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Not really. Switching to what we have elsewhere.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 8 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Sure.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 9 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Will keep it as is.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 10 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Added a todo for that.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 11 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Done.

File: core/src/main/java/org/elasticsearch/search/sort/ParameterParser.java (hunk 12 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/ParameterParser.java]: I tried to keep this method out of the SortBuilder intentionally so as not to have to add this method to every class implementing the SortBuilder interface, not sure if that makes sense? Also should have added a comment to potentially move this into the sortbuilder interface once done.

File: core/src/test/java/org/elasticsearch/search/sort/AbstractSearchSourceItemTestCase.java (hunk 14 above)
Comment (MaineC) [inline core/src/test/java/org/elasticsearch/search/sort/AbstractSearchSourceItemTestCase.java]: Renamed, also removed the todo as it was a leftover.

File: core/src/test/java/org/elasticsearch/search/sort/AbstractSearchSourceItemTestCase.java (hunk 13 above)
Comment (MaineC) [inline core/src/test/java/org/elasticsearch/search/sort/AbstractSearchSourceItemTestCase.java]: Leftover, fixed, thanks for pointing it out.

**Comment by MaineC** [comment]
@cbuescher Went over your comments and tried to address as many as possible. Left a few questions as well.

Does it make sense to merge this and do the switchover to a to be implemented build method in a separate PR?

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 15]
```diff
@@ -22,41 +22,109 @@
...
     /**
      * Constructs a new distance based sort on a geo point like field.
      *
      * @param fieldName The geo point like field name.
+     * @param points The points to create the range distance facets from.
      */
-    public GeoDistanceSortBuilder(String fieldName) {
+    public GeoDistanceSortBuilder(String fieldName, GeoPoint... points) {
         this.fieldName = fieldName;
+        this.points.addAll(Arrays.asList(points));
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Maybe you could put the validation removed from toCContent here. (point.size > 0)

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 16]
```diff
@@ -193,14 +320,200 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public boolean equals(Object object) {
+        if (super.equals(object)) {
+            return true;
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Nit: I might get something wrong, but seems trappy to me since it goes to Object equals now (and does the right thing), but if any other superclass (like SortBuilder) would implements it's own equals() this would short-circuit everything. Of course there are tests for it now, but I'd do an explicit check for reference equality here.

File: core/src/main/java/org/elasticsearch/search/sort/ParameterParser.java (hunk 12 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/ParameterParser.java]: +1 on either making the name more specific or merging this with SortBuilder at the end.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 17]
```diff
@@ -22,41 +22,109 @@
...
     private final List<GeoPoint> points = new ArrayList<>();
-    private final List<String> geohashes = new ArrayList<>();
 
-    private GeoDistance geoDistance;
-    private DistanceUnit unit;
-    private SortOrder order;
-    private String sortMode;
+    private GeoDistance geoDistance = GeoDistance.DEFAULT;
+    private DistanceUnit unit = DistanceUnit.DEFAULT;
+    private SortOrder order = SortOrder.DEFAULT;
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Can you re-check the default? I think it should be "ASC" since GeoDistanceSortParser used to set `reverse = false` by default to far, which I think translates to ASC?

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 18]
```diff
@@ -147,42 +265,51 @@ public GeoDistanceSortBuilder setNestedPath(String nestedPath) {
...
 
-        if (unit != null) {
-            builder.field("unit", unit);
-        }
-        if (geoDistance != null) {
-            builder.field("distance_type", geoDistance.name().toLowerCase(Locale.ROOT));
-        }
+        builder.field("unit", unit);
+        builder.field("distance_type", geoDistance.name().toLowerCase(Locale.ROOT));
         if (order == SortOrder.DESC) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: To be on the save side and not rely on the parser picking the right default when "reverse" is not set, maybe we should also write the `reverse` field for ASC?

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 17 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Hmm, doc says `The order defaults to desc when sorting on the _score, and defaults to asc when sorting on anything else.`, so maybe having the default in the enum is missleading.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 1 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: I see, so parser always sets both "order" and "mode", regardless of whether they are set by the user. But what if we only go through the java api, use a plain builder and set "reverse = false". Translated to json this should give us "mode = MIN", but only if not explicitely set by the user otherwise, no? Sorry, haven't got a good solution myself so far either.

**Comment by cbuescher** [comment]
@MaineC Thanks, I went through the changes, left some small remarks and idea about relation between the "mode" and "order" parmeter but this is indeed a tricky detail. Maybe setting the default is the right thing here, I'm still wondering what we are then missing on the builder side though.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 1 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Decided to not do any of this validation/setting while parsing xcontent but postpone to build method implementation when discussing f2f

File: core/src/main/java/org/elasticsearch/search/sort/ParameterParser.java (hunk 12 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/ParameterParser.java]: Renamed, added a comment, will open an issue to move to SortBuilder once this is merged.

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 16 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: But then, if any of the superclasses did implement their own equals - wouldn't I actually want to check their equals implementation?

Can add the check for reference equality explicitly of course.

Somehow this discussion reminds me of https://en.wikipedia.org/wiki/Liskov_substitution_principle

Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Makes sense, not nitpicking at all. (after re-reading the code twice...)

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java (hunk 17 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Thanks for double checking the docs. Changed. Tricky indeed.

**Comment by MaineC** [comment]
@cbuescher Made a few changes, feel free to have another look.

**Comment by cbuescher** [comment]
Thanks, LGTM
